File: LeetCode/Python/Easy/13-Roman-To-Integer/solution.py
class Solution:
    def romanToInt(self, s: str) -> int:
        result: int=0

        n: int=len(s)

        numerals=["I", "V", "X", "L", "C", "D", "M", "IX", "IV", "XL", "XC", "CD", "CM"]

        hashmap=self.populate_hashmap(numerals)

        index: int=0

        while index < n:
            key: str=str(s[index])
            
            if s[index]=='I':
                if index + 1 < n and (s[index + 1]=='V' or s[index + 1]=='X'):
                    key=key + s[index + 1]

                    result+=hashmap[key]

                    index+=2
                else:
                    result+=hashmap[key]

                    index+=1

            elif s[index]=='X':
                if index + 1 < n and (s[index + 1]=='L' or s[index + 1]=='C'):
                    key=key + s[index + 1]

                    result+=hashmap[key]

                    index+=2
                else:
                    result+=hashmap[key]

                    index+=1
                
            elif s[index]=='C':
                if index + 1 < n and (s[index + 1]=='D' or s[index + 1]=='M'):
                    key=key + s[index + 1]

                    result+=hashmap[key]

                    index+=2
                else:
                    result+=hashmap[key]

                    index+=1
            else:
                result+=hashmap[key]

                index+=1
        
        return result


    def populate_hashmap(self, numerals: List[str]) -> dict[str, int]:
        hashmap: dict[str, int]=dict()

        for numeral in numerals:
            if numeral=="I":
                hashmap[numeral]=1

            elif numeral=="IV":
                hashmap[numeral]=4

            elif numeral=="V":
                hashmap[numeral]=5

            elif numeral=="IX":
                hashmap[numeral]=9

            elif numeral=="X":
                hashmap[numeral]=10

            elif numeral=="XL":
                hashmap[numeral]=40

            elif numeral=="L":
                hashmap[numeral]=50

            elif numeral=="XC":
                hashmap[numeral]=90

            elif numeral=="C":
                hashmap[numeral]=100

            elif numeral=="CD":
                hashmap[numeral]=400

            elif numeral=="D":
                hashmap[numeral]=500

            elif numeral=="CM":
                hashmap[numeral]=900

            elif numeral=="M":
                hashmap[numeral]=1000

        return hashmap

File: LeetCode/Rust/Easy/13-Roman-To-Integer/solution.rs
use std::collections::HashMap;

impl Solution
{
    pub fn roman_to_int(s: String) -> i32
    {
        let mut result: i32=0;

        let n: usize=s.len();

        let hashmap: HashMap<String, u32>=Solution::populate_hashmap();

        let mut index: usize=0;

        while index < n
        {
            let mut character: char=s.chars().nth(index).unwrap();

            let mut key: String=String::from(character);

            match(character)
            {
                'I' =>
                {
                    if index + 1 < n
                    {
                        let next: char=s.chars().nth(index + 1).unwrap();
                        
                        if next=='V'|| next=='X'
                        {
                            key.push(next);

                            index+=2;
                        }
                        else
                        {
                            index+=1;
                        }
                        
                        let value: &u32=hashmap.get(&key).unwrap();

                        let amount: i32=*value as i32;

                        result+=amount;
                    }
                    else
                    {
                        let value: &u32=hashmap.get(&key).unwrap();

                        let amount: i32=*value as i32;

                        result+=amount;

                        index+=1
                    }
                },

                'X' =>
                {
                    if index + 1 < n
                    {   
                        let next: char=s.chars().nth(index + 1).unwrap();
                        
                        if next=='L'|| next=='C'
                        {
                            key.push(next);

                            index+=2;
                        }
                        else
                        {
                            index+=1;
                        }
                        
                        let value: &u32=hashmap.get(&key).unwrap();

                        let amount: i32=*value as i32;

                        result+=amount;
                    }
                    else
                    {
                        let value: &u32=hashmap.get(&key).unwrap();

                        let amount: i32=*value as i32;

                        result+=amount;

                        index+=1
                    }
                },

                'C' =>
                {
                    if index + 1 < n
                    {
                        let next: char=s.chars().nth(index + 1).unwrap();
                        
                        if next=='D'|| next=='M'
                        {
                            key.push(next);

                            index+=2;
                        }
                        else
                        {
                            index+=1;
                        }
                        
                        let value: &u32=hashmap.get(&key).unwrap();

                        let amount: i32=*value as i32;

                        result+=amount;
                    }
                    else
                    {
                        let value: &u32=hashmap.get(&key).unwrap();

                        let amount: i32=*value as i32;

                        result+=amount;

                        index+=1
                    }
                },

                _ =>
                {
                    let value: &u32=hashmap.get(&key).unwrap();

                    let amount: i32=*value as i32;

                    result+=amount;

                    index+=1
                }
            };
        }

        result
    }

    pub fn populate_hashmap() -> HashMap<String, u32>
    {
        let symbols: Vec<String>=vec!["I".to_string(), "IV".to_string(), "V".to_string(), "IX".to_string(), "X".to_string(), "XL".to_string(), "L".to_string(), "XC".to_string(), "C".to_string(), "CD".to_string(), "D".to_string(), "CM".to_string(), "M".to_string()];

        let mut hashmap: HashMap<String, u32>=HashMap::new();

        for symbol in symbols
        {
            match(symbol.as_str())
            {
                "I" =>
                {
                    hashmap.insert(symbol, 1);
                },

                "IV" =>
                {
                    hashmap.insert(symbol, 4);
                },

                "V" =>
                {
                    hashmap.insert(symbol, 5);
                },

                "IX" =>
                {
                    hashmap.insert(symbol, 9);
                },

                "X" =>
                {
                    hashmap.insert(symbol, 10);
                },

                "XL" =>
                {
                    hashmap.insert(symbol, 40);
                },

                "L" =>
                {
                    hashmap.insert(symbol, 50);
                },

                "XC" =>
                {
                    hashmap.insert(symbol, 90);
                },

                "C" =>
                {
                    hashmap.insert(symbol, 100);
                },

                "CD" =>
                {
                    hashmap.insert(symbol, 400);
                },

                "D" =>
                {
                    hashmap.insert(symbol, 500);
                },

                "CM" =>
                {
                    hashmap.insert(symbol, 900);
                },

                "M" =>
                {
                    hashmap.insert(symbol, 1000);
                },

                _ => (),
            };
        }

        hashmap
    }
}

impl Solution
{
    pub fn roman_to_int(s: String) -> i32
    {
        let mut result: i32=0;

        let n: usize=s.len();

        let translate=|character: char| match(character)
        {
            'I' => 1,
            'V' => 5,
            'X' => 10,
            'L' => 50,
            'C' => 100,
            'D' => 500,
            'M' => 1000,
            _ => 0
        };

        let mut values: Vec<i32>=s.chars().map(translate).collect();

        for index in 0..n-1
        {
            if values[index] < values[index + 1]
            {
                values[index]*=-1;
            }
        }

        result=values.iter().sum();

        result
    }
}